src/cipher/mod.rs

```diff
@@ -0,0 +1,359 @@
+
+use std::io;
+
+/// Maximum number of scalars allowed per message
+pub const MESSAGE_CAPACITY: usize = 2;
+
+/// Number of scalars used in a cipher
+pub const CIPHER_SIZE: usize = MESSAGE_CAPACITY + 1;
+
+/// Bytes consumed on serialization of the poseidon cipher
+pub const ENCRYPTED_DATA_SIZE: usize = CIPHER_SIZE * 32;
+
+/// Plonk gadget for Poseidon encryption
+pub mod gadget;
+
+/// Encapsulates an encrypted data
+///
+/// This implementation is optimized for a message containing 2 scalars
+///
+/// # Examples
+/// ```
+/// use dusk_plonk::jubjub::{dhke, ExtendedPoint, GENERATOR};
+/// use dusk_plonk::prelude::*;
+/// use poseidon252::cipher::{PoseidonCipher, MESSAGE_CAPACITY};
+///
+/// use std::ops::Mul;
+///
+/// fn main() {
+///     let mut rng = rand::thread_rng();
+///
+///     // Generate a secret and a public key for Bob
+///     let bob_secret = JubJubScalar::random(&mut rng);
+///     let bob_public = GENERATOR.to_niels().mul(&bob_secret);
+///
+///     // Generate a secret and a public key for Alice
+///     let alice_secret = JubJubScalar::random(&mut rng);
+///     let alice_public = GENERATOR.to_niels().mul(&alice_secret);
+///
+///     // Generate a secret message
+///     let a = BlsScalar::random(&mut rng);
+///     let b = BlsScalar::random(&mut rng);
+///     let message = [a, b];
+///
+///     // Bob's view (sender)
+///     // The cipher and nonce are safe to be broadcasted publicly
+///     let (cipher, nonce) = sender(&bob_secret, &alice_public, &message);
+///
+///     // Alice's view (receiver)
+///     let decrypted_message =
+///         receiver(&alice_secret, &bob_public, &cipher, &nonce);
+///
+///     // Successful communication
+///     assert_eq!(decrypted_message, message);
+/// }
+///
+/// fn sender(
+///     sender_secret: &JubJubScalar,
+///     receiver_public: &ExtendedPoint,
+///     message: &[BlsScalar],
+/// ) -> (PoseidonCipher, BlsScalar) {
+///     // Use the Diffie-Hellman protocol to generate a shared secret
+///     let shared_secret = dhke(sender_secret, receiver_public);
+///
+///     // Generate a random nonce that will be public
+///     let nonce = BlsScalar::random(&mut rand::thread_rng());
+///
+///     // Encrypt the message
+///     let cipher = PoseidonCipher::encrypt(&message, &shared_secret, &nonce);
+///
+///     (cipher, nonce)
+/// }
+///
+/// fn receiver(
+///     receiver_secret: &JubJubScalar,
+///     sender_public: &ExtendedPoint,
+///     cipher: &PoseidonCipher,
+///     nonce: &BlsScalar,
+/// ) -> [BlsScalar; MESSAGE_CAPACITY] {
+///     // Use the Diffie-Hellman protocol to generate a shared secret
+///     let shared_secret = dhke(receiver_secret, sender_public);
+///
+///     // Decrypt the message
+///     cipher
+///         .decrypt(&shared_secret, &nonce)
+///         .expect("Failed to decrypt!")
+/// }
+/// ```
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Ord, PartialOrd, Default)]
+pub struct PoseidonCipher {
+    cipher: [BlsScalar; CIPHER_SIZE],
+}
```

Comment:
Does it need to be a "new type"? Since if we could avoid that it would be nice (less stuff to pass through GRPC/ProtoBuff/WASM...)

Reply:
This structure is prone to change depending on the encryption parameters we take.

It is better we encapsulate it as much as possible to avoid high-impact changes in the future.

Reply:
LGTM